### PR TITLE
(WIP) feat: HealthCheck

### PR DESCRIPTION
**Pull request by Juliens**

### CODE REVIEW

File: server.go [hunk 1]
```diff
@@ -524,6 +525,9 @@ func (server *Server) loadConfig(configurations configs, globalConfiguration Glo
 	redirectHandlers := make(map[string]http.Handler)
 
 	backends := map[string]http.Handler{}
+
+	backendsHealcheck := map[string]*healthcheck.BackendHealthCheck{}
```
Comment (emilevauge) [inline server.go]: typo

File: healthcheck/healthcheck.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+var singleton *HealthCheck
+var once sync.Once
+
+// GetHealthCheck Get HealtchCheck Singleton
+func GetHealthCheck() *HealthCheck {
+	once.Do(func() {
+		singleton = newHealthCheck()
+		singleton.execute()
+	})
+	return singleton
```
Comment (emilevauge) [inline healthcheck/healthcheck.go]: Maybe an `init` function would be simpler ?

File: healthcheck/healthcheck.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+						backend.DisabledURLs = append(backend.DisabledURLs, url)
+					}
+				}
+			}
+		}
+	}()
+}
+
+func testHealth(serverURL *url.URL, checkURL string) bool {
+	resp, err := http.Get(serverURL.String() + checkURL)
```
Comment (emilevauge) [inline healthcheck/healthcheck.go]: We need a timeout here

File: types/types.go [hunk 4]
```diff
@@ -35,6 +36,11 @@ type CircuitBreaker struct {
 	Expression string `json:"expression,omitempty"`
 }
 
+// HealthCheck holds healthchk configuration
```
Comment (emilevauge) [inline types/types.go]: >HealthCheck holds healthchk 

*HealthCheck

File: healthcheck/healthcheck.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+func (hc *HealthCheck) execute(ctx context.Context) {
+	for backendID, backend := range hc.Backends {
+		go func(backendID string, backend *BackendHealthCheck) {
+			for {
+				ticker := time.NewTicker(time.Second * 30)
+				select {
+				case <-ctx.Done():
+					log.Debugf("Refreshing All Healthcheck goroutines")
```
Comment (emilevauge) [inline healthcheck/healthcheck.go]: You mean stopping ?

File: healthcheck/healthcheck.go [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+					return
+				case <-ticker.C:
+					log.Debugf("Refreshing Healthcheck for backend %s ", backendID)
+					enabledURLs := backend.lb.Servers()
+					var newDisabledURLs []*url.URL
+					for _, url := range backend.DisabledURLs {
+						if testHealth(url, backend.URL) {
+							log.Debugf("Upsert %s", url.String())
```
Comment (emilevauge) [inline healthcheck/healthcheck.go]: Could you log more info in this log ?

File: healthcheck/healthcheck.go [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+							backend.lb.UpsertServer(url, roundrobin.Weight(1))
+						} else {
+							newDisabledURLs = append(newDisabledURLs, url)
+						}
+					}
+					backend.DisabledURLs = newDisabledURLs
+
+					for _, url := range enabledURLs {
+						if !testHealth(url, backend.URL) {
+							log.Debugf("Remove %s", url.String())
```
Comment (emilevauge) [inline healthcheck/healthcheck.go]: Could you log more info in this log ?

**Comment by bakins** [comment]
@Juliens are you still working on this? I'm interested in this feature and can take a look at addressing the comments, etc, if you don't have time.

**Comment by Juliens** [comment]
@bakins I think that will be difficult to handle this before the next release (tomorrow, I guess). I'll try to do it before the end of the week.

**Comment by klausenbusk** [comment]
Seems like I can close my issue then :) Good work.
https://github.com/containous/traefik/issues/824

**Pull request merged**